File: Restaurant/src/main/java/com/finalproject/Restaurant/model/Activity.java
package com.finalproject.Restaurant.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.*;

@Entity
@Table(name = "Activity")
@Data
@AllArgsConstructor
@NoArgsConstructor
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
public class Activity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer activityId;

    @Column(nullable = false, length = 255)
    private String activityName;

    @Column(columnDefinition = "TEXT")
    private String descriptionActivity;

    @Column(nullable = false)
    private Date inviteDate;

    @Column(nullable = false)
    private Date postDate;

    @Column(nullable = false, length = 50)
    private String statusPost;

    @Column(nullable = false)
    private Boolean isOwnerSelect;

    @ManyToMany(cascade = CascadeType.ALL)
    @JoinTable(
            name = "member_join_activity",
            joinColumns = @JoinColumn(name = "activityId"),
            inverseJoinColumns = @JoinColumn(name = "activityMemberId")
    )
    private List<ActivityMember> activityMembers = new ArrayList<>();

    @ManyToOne
    @JoinColumn(name = "restaurantTypeId")
    private RestaurantType restaurantType;

    @ManyToOne
    @JoinColumn(name = "restaurantId")
    private Restaurant restaurant;
}

File: ProjectApplication/lib/model/Activity.dart
import 'package:newproject/model/ActivityMember.dart';
import 'package:newproject/model/RestaurantType.dart';
import 'Restaurant.dart';

class Activity {
  final int? activityId;
  final String? activityName;
  final String? descriptionActivity;
  final DateTime? inviteDate;
  final DateTime? postDate;
  final String? statusPost;
  final bool? isOwnerSelect;
  final RestaurantType? restaurantType;
  final Restaurant? restaurant;
  final List<ActivityMember> activityMembers;

  Activity({
    this.activityName,
    this.descriptionActivity,
    this.inviteDate,
    this.postDate,
    this.statusPost,
    this.isOwnerSelect,
    this.restaurantType,
    this.restaurant,
    this.activityMembers = const [],
    this.activityId,
  });

  factory Activity.fromJson(Map<String, dynamic> json) => Activity(
        activityId: json['activityId'],
        activityName: json['activityName'],
        descriptionActivity: json['descriptionActivity'],
        inviteDate: _parseToLocal(json['inviteDate'] ?? ''),
        postDate: _parseToLocal(json['postDate'] ?? ''),
        statusPost: json['statusPost'],
        isOwnerSelect: json['isOwnerSelect'],
        restaurantType: json['restaurantType'] != null
            ? RestaurantType.fromJson(json['restaurantType'])
            : null,
        restaurant: json['restaurant'] != null
            ? Restaurant.fromRestaurantJson(json['restaurant'])
            : null,
        activityMembers: (json['activityMembers'] as List<dynamic>?)
                ?.map((e) => ActivityMember.fromJson(e))
                .toList() ??
            [],
      );

  Map<String, dynamic> toJson() => {
        "activityName": activityName,
        "descriptionActivity": descriptionActivity,
        "inviteDate": inviteDate?.toUtc().toIso8601String(),
        "postDate": postDate?.toUtc().toIso8601String(),
        "statusPost": statusPost,
        "isOwnerSelect": isOwnerSelect,

        // ส่งต่อเมื่อรู้ id เท่านั้น (อย่าส่ง object ว่าง)
        if (restaurantType?.restaurantTypeId != null)
          "restaurantType": {"restaurantTypeId": restaurantType!.restaurantTypeId},

        if (restaurant?.restaurantId != null)
          "restaurant": {"restaurantId": restaurant!.restaurantId},

        'activityMembers': activityMembers.map((e) => e.toJson()).toList(),
      };

  static DateTime? _parseToLocal(dynamic v) {
    if (v == null) return null;
    final d = DateTime.tryParse(v.toString());
    if (d == null) return null;
    return d.isUtc ? d.toLocal() : d;
  }
}
